juju-tutorials/how-to-deploy-openstakc-on-openstack.md

```diff
@@ -0,0 +1,371 @@
+3. OpenStack is configured with valid credentials (e.g. `source novarc`, represent the OpenStack credential, 
+   belonging to OpenStack on top of which you are deploying [OpenStack-base][openstack-base] and if you don't have it, 
+   ask your OpenStack administrator)
+
+### OpenStack
+
+Deploy the [OpenStack-base][openstack-base] bundle with the custom overlay.
+
+##### Memory overlay
+```yaml
+applications:
+  ceph-osd:
+    storage:
+      osd-devices: 40GB
+  nova-compute:
+    annotations:
+    constraints: cores=6 mem=8G
+```
+
+##### Deployment
+
+```bash
+git clone https://github.com/openstack-charmers/openstack-bundles.git
+juju deploy ./openstack-bundles/stable/openstack-base/bundle.yaml --overlay ./openstack-bundles/stable/overlays/openstack-base-virt-overlay.yaml --overlay ./resource-overlay.yaml
+```
+
+##### Initialize and unseal the Vault and create a self-signed root CA
+
+1. Wait for the `vault/0` unit to be in the `blocked` state with message `Vault needs to be initialized`.
+   ```bash
+   juju ssh vault/0
+   ```
+2. Initialize the vault deployment.
+   ```bash
+   export VAULT_ADDR="http://127.0.0.1:8200"
+   vault operator init -key-shares=3 -key-threshold=3 > vault-keys
+   ```
+4. Unseal all keys and generate root CA.
+   ```bash
+   for key in $(grep -E "Unseal Key .*: " vault-keys | cut -c15-)
+   do 
+       vault operator unseal $key
+   done
+   ```
+5. Initial TTL to associate with the token.
+   ```bash
+   export VAULT_TOKEN=$(grep -E "Initial Root Token: " vault-keys | cut -c21-)
+   vault token create -ttl=10m
+   ```
+6. Detach from the `vault/0` unit. 
+5. Authorize the vault unit.
+   ```bash
+   export VAULT_TOKEN=$(juju ssh vault/0 'grep -E "Initial Root Token: " vault-keys | cut -c21-')
+   juju run-action --wait vault/leader authorize-charm token=$VAULT_TOKEN
+   juju run-action --wait vault/leader generate-root-ca
+   ```
+   
+##### Add external L3 connectivity
+
+This section deals with the prerequisites for network configuration and is required for the full functionality of 
+the deployed OpenStack. It guarantees a connection between individual units.
+
+[note status="OpenStack credentials"]
+Don't forget to get credentials for the main OpenStack. (e.g. `source novarc`, look at "Requirements" section)
+[/note]
+
+1. Get OpenStack admin net.
+   ```bash
+   export ADMIN_NET=$(openstack network list | grep admin_net | cut -d"|" -f3 | tr -d " ")
+   ```
+2. Create new port for each nova-compute unit. (by default there are 3 units)
+   ```bash
+   unset BRIDGE_INTERFACE_MAPPING  # clear value
+
+   for i in 0 1 2
+   do
+       # create new port
+       export NEW_PORT_$i_ID=$(openstack port create --network $ADMIN_NET data-port | grep ' id ' | cut -d"|" -f3 | tr -d " ")
+       echo "  create new port with id $NEW_PORT_$i_ID"
+    
+       # get ovn chassis ip
+       export OVN_CHASSIS_$i_IP=$(juju status | grep ovn-chassis/$i | tr -s " " | cut -d" " -f5)
```

Comment:
I'm testing it now, if the unit is blocked, it doesn't work.
```bash
➜  NovaComputeFixVolumeAttach juju status --format json | jq -r '.applications."ovn-chassis"'
{
  "charm": "cs:ovn-chassis-6",
  "series": "focal",
  "os": "ubuntu",
  "charm-origin": "jujucharms",
  "charm-name": "ovn-chassis",
  "charm-rev": 6,
  "charm-version": "404a44d",
  "charm-profile": "juju-k8-ovn-chassis-6",
  "exposed": false,
  "application-status": {
    "current": "error",
    "message": "hook failed: \"ovsdb-relation-changed\"",
    "since": "28 Jan 2021 13:20:25+01:00"
  },
  "relations": {
    "certificates": [
      "vault"
    ],
    "nova-compute": [
      "nova-compute"
    ],
    "ovsdb": [
      "ovn-central"
    ]
  },
  "subordinate-to": [
    "nova-compute"
  ],
  "version": "20.03.1",
  "endpoint-bindings": {
    "": "alpha",
    "amqp": "alpha",
    "certificates": "alpha",
    "data": "alpha",
    "juju-info": "alpha",
    "nova-compute": "alpha",
    "nrpe-external-master": "alpha",
    "ovsdb": "alpha",
    "ovsdb-subordinate": "alpha"
  }
}
```